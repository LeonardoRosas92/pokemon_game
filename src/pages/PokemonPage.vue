<template>
  <h1 class="font-pokemon">¿Who's that PokéMoN?</h1>
  <div v-if="!pokemon">
    <Loader />
  </div>
  <div v-else>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/>
    <template v-if="showAnswer">
      <div class="result">
        <h2>{{ message }}</h2>
      </div>
    </template>
  </div>
  <PokemonScore :score="score" />
  <NewGame v-if="showModal" :score="score" :message="message" @newGame="newGame"/>
  <Lives :lives="lives" />
</template>

<script>
import PokemonPicture from "../components/PokemonPicture.vue";
import PokemonOptions from "../components/PokemonOptions.vue";
import PokemonScore from "../components/PokemonScore.vue";
import Loader from "../components/Loader.vue";
import NewGame from "../components/NewGame.vue";
import Lives from "../components/Lives.vue"

import getPokemonOptions from "@/helpers/getPokemonOptions";

//getPokemonOptions();
export default {
  name: "PokemonPage",
  components: {
    PokemonPicture,
    PokemonOptions,
    PokemonScore,
    Loader,
    NewGame,
    Lives
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      showModal: false,
      message: "",
      score: 0,
      lives: 3
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      const randomInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[randomInt];
    },
    nextPokemon() {
      console.log("NextPokemon");
      setTimeout(() => {
        this.showPokemon = false;
        this.showAnswer = false;
        this.mixPokemonArray();
      }, 3000);
    },
    checkAnswer(pokemonId) {
      this.showPokemon = true;
      if (pokemonId === this.pokemon.id) {
        this.showAnswer = true;
        this.message = `Correct, it's ${this.pokemon.name}`;
        this.score = this.score + Math.floor(Math.random() * 20);
        this.nextPokemon();
      } else {
        console.log(this.lives);
        this.lives = this.lives - 1;
        if (!this.lives > 0) {
          this.showModal = true;
          this.message = `Oops, it is ${this.pokemon.name}`;
        }else{
          this.nextPokemon();
        }
        
        
      }
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArr = [];
      this.pokemon = null;
      this.showModal = false;
      this.score = 0;
      this.lives = 3;
      this.mixPokemonArray();
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style scoped>
</style>