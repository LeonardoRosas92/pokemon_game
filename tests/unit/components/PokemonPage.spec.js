import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'
describe('PokemonPage component', () => {

    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage);
    });

    test('Debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('Debe de llamar mixPokemonArray al montar', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
        const wrapper = shallowMount(PokemonPage);
        expect(mixPokemonArraySpy).toHaveBeenCalled();
    });

    test('Debe de hacer match con el snapshot cuando cargan los pokemos', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showModal: false,
                    message: "",
                    score: 0,
                    lives: 3
                }
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showModal: false,
                    message: "",
                    score: 0,
                    lives: 3
                }
            }
        });

        const picture = wrapper.find('pokemon-picture-stub');
        const options = wrapper.find('pokemon-options-stub');
        //PokemonPicture debe de existir 
        expect(picture.exists()).toBeTruthy();
        //PokemonOptions debe de existir
        expect(options.exists()).toBeTruthy();
        //PokemonPicture attribute pokemon === '1'
        expect(picture.attributes('pokemonid')).toBe('1');

        expect(options.attributes('pokemons')).toBeTruthy();
    });

    test('Pruebas con checkAnswer', async() => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showModal: false,
                    message: "",
                    score: 0,
                    lives: 3
                }
            }
        });

        await wrapper.vm.checkAnswer(1);
        expect(wrapper.find('h2').exists()).toBeTruthy();
        expect(wrapper.vm.showPokemon).toBe(true);
        expect(wrapper.find('h2').text()).toBe(`Correct, it's ${pokemons[0].name}`);
    });
})