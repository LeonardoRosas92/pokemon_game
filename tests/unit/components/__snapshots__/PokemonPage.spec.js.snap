// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component Debe de hacer match con el snapshot 1`] = `
<h1 class="font-pokemon">¿Who's that PokéMoN?</h1>
<div>
  <loader-stub></loader-stub>
</div>
<pokemon-score-stub score="0" message="Your Score"></pokemon-score-stub>
<!--v-if-->
<lives-stub lives="3"></lives-stub>
`;

exports[`PokemonPage component Debe de hacer match con el snapshot cuando cargan los pokemos 1`] = `
<h1 class="font-pokemon">¿Who's that PokéMoN?</h1>
<div>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
<pokemon-score-stub score="0" message="Your Score"></pokemon-score-stub>
<!--v-if-->
<lives-stub lives="3"></lives-stub>
`;
